examples/miral-shell/decoration/decoration.cpp

```diff
@@ -0,0 +1,474 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "decoration.h"
+
+#include "mir/geometry/dimensions.h"
+#include "mir/geometry/forward.h"
+#include "mir/geometry/rectangle.h"
+#include "mir_toolkit/common.h"
+#include "mir_toolkit/cursors.h"
+
+#include "miral/decoration_adapter.h"
+#include "miral/decoration.h"
+#include "miral/decoration_window_state.h"
+#include "miral/decoration_basic_manager.h"
+
+#include <memory>
+
+namespace geom = mir::geometry;
+namespace md = miral::decoration;
+
+UserDecoration::UserDecoration() = default;
+
+auto UserDecoration::InputManager::buttons() const -> std::vector<std::shared_ptr<Widget const>>
+{
+    std::vector<std::shared_ptr<Widget const>> button_rects;
+    for (auto const& widget : widgets)
+    {
+        if (widget->button)
+            button_rects.push_back(widget);
+    }
+
+    return button_rects;
```

Comment:
Would've been nice to do this using the ranges API but it seems not ready yet...

Reply:
"not ready yet"? There's not much code using ranges in our codebase, but we have used it

Reply:
Looks like I'm who's not ready. I spent about 20 minutes trying to filter and collect but couldn't figure it out.